---
title: "4_generate_gene_tables"
author: "Matt George; mngeorge@uw.edu"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = TRUE, message = FALSE, cache = TRUE)
```

```{bash, engine.path="C:/Windows/System32/bash.exe"}
echo $PATH

mkdir gene_tables/
```

# Need to detach bioconductor (from last script) for this script to work
```{r}
#Detach all  packages
detachAllPackages <- function() {

  basic.packages <- c("package:stats","package:graphics","package:grDevices","package:utils","package:datasets","package:methods","package:base")

  package.list <- search()[ifelse(unlist(gregexpr("package:",search()))==1,TRUE,FALSE)]

  package.list <- setdiff(package.list,basic.packages)

  if (length(package.list)>0)  for (package in package.list) detach(package, character.only=TRUE)

}

detachAllPackages()
```

# Make file structure
```{bash engine.path="C:/Windows/System32/bash.exe", include=FALSE}

mkdir blast/
mkdir GFF/

```

### Load libraries
```{r load_libraries, inlcude = TRUE}

## clear
rm(list=ls())

# List of packages we want to install (run every time)
load.lib<-c("RColorBrewer","readxl","ggpubr","tidyverse","tibble","stringr","beepr","gplots")

# Select only the packages that aren't currently installed (run every time)
install.lib <- load.lib[!load.lib %in% installed.packages()]

# And finally we install the missing packages, including their dependency.
for(lib in install.lib) install.packages(lib,dependencies=TRUE)
# After the installation process completes, we load all packages.
sapply(load.lib,require,character=TRUE)

                        
```

#Load data
```{r load_data, inlcude = TRUE}

# Load LOCID to gene description tables, compare, find gigas and mito genes
control_ploidy <- read.delim("DESEQ2_output/control_ploidy/CONTROL_PLOIDY-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
control_ploidy <- control_ploidy %>% select(gene, padj, log2FoldChange, description)

control_ploidy_bg <- read.delim("DESEQ2_output/control_ploidy/CONTROL_PLOIDY-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
control_ploidy_bg <- control_ploidy_bg %>% select(gene, padj, log2FoldChange, description)

heat_ploidy <- read.delim("DESEQ2_output/heat_ploidy/HEAT_PLOIDY-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
heat_ploidy <- heat_ploidy %>% select(gene, padj, log2FoldChange, description)

heat_ploidy_bg <- read.delim("DESEQ2_output/heat_ploidy/HEAT_PLOIDY-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
heat_ploidy_bg <- heat_ploidy_bg %>% select(gene, padj, log2FoldChange, description)

desiccation_ploidy <- read.delim("DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
desiccation_ploidy <- desiccation_ploidy %>% select(gene, padj, log2FoldChange, description)

desiccation_ploidy_bg <- read.delim("DESEQ2_output/desiccation_ploidy/DESICCATION_PLOIDY-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
desiccation_ploidy_bg <- desiccation_ploidy_bg %>% select(gene, padj, log2FoldChange, description)

diploid_heat <- read.delim("DESEQ2_output/diploid_heat/DIPLOID_HEAT-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
diploid_heat <- diploid_heat %>% select(gene, padj, log2FoldChange, description)

diploid_heat_bg <- read.delim("DESEQ2_output/diploid_heat/DIPLOID_HEAT-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
diploid_heat_bg <- diploid_heat_bg %>% select(gene, padj, log2FoldChange, description)

diploid_desiccation <- read.delim("DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
diploid_desiccation <- diploid_desiccation %>% select(gene, padj, log2FoldChange, description)

diploid_desiccation_bg <- read.delim("DESEQ2_output/diploid_desiccation/DIPLOID_DESICCATION-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
diploid_desiccation_bg <- diploid_desiccation_bg %>% select(gene, padj, log2FoldChange, description)

triploid_heat <- read.delim("DESEQ2_output/triploid_heat/TRIPLOID_HEAT-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
triploid_heat <- triploid_heat %>% select(gene, padj, log2FoldChange, description)

triploid_heat_bg <- read.delim("DESEQ2_output/triploid_heat/TRIPLOID_HEAT-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
triploid_heat_bg <- triploid_heat_bg %>% select(gene, padj, log2FoldChange, description)

triploid_desiccation <- read.delim("DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-SIG-DEG-apeglm.csv", sep = " ", header = TRUE)
triploid_desiccation <- triploid_desiccation %>% select(gene, padj, log2FoldChange, description)

triploid_desiccation_bg <- read.delim("DESEQ2_output/triploid_desiccation/TRIPLOID_DESICCATION-ALL-DEG-apeglm.csv", sep = " ", header = TRUE)
triploid_desiccation_bg <- triploid_desiccation_bg %>% select(gene, padj, log2FoldChange, description)

diploid_MS_unique <- read.delim("DESEQ2_output/DEG_diploid_multi_unique.csv", sep = " ", header = TRUE)
diploid_MS_unique <- diploid_MS_unique %>% select(gene, padj, log2FoldChange, description)

triploid_MS_unique <- read.delim("DESEQ2_output/DEG_triploid_multi_unique.csv", sep = " ", header = TRUE)
triploid_MS_unique <- triploid_MS_unique %>% select(gene, padj, log2FoldChange, description)


```


# Download blast results and genome features from gannet 
```{bash engine.path="C:/Windows/System32/bash.exe", include=FALSE}

#wget -r \
#--no-directories --no-parent \
#-P . \
#-A blast/20210605-cgigas-roslin-mito-blastx.outfmt6 https://gannet.fish.washington.edu/panopea/NOPP-gigas-ploidy-temp/20210605-blast/ \
#--no-check-certificate

#wget -r \
#--no-directories --no-parent \
#-P . \
#-A blast/uniprot-SP-GO.sorted https://gannet.fish.washington.edu/panopea/NOPP-gigas-ploidy-temp/20210605-blast/ \
#--no-check-certificate

wget -r \
--no-directories --no-parent \
-P . \
-A GFF/cgigas_uk_roslin_v1_mRNA.gff https://gannet.fish.washington.edu/panopea/NOPP-gigas-ploidy-temp/20220908-GFF/ \
--no-check-certificate

```

# Generate Gene2GO list (all results stored at: https://gannet.fish.washington.edu/panopea/NOPP-gigas-ploidy-temp/20220908-GOterm-Annotation/)
```{bash, engine.path="C:/Windows/System32/bash.exe"}

# Check output
#head 20210605-cgigas-roslin-mito-blastx.outfmt6
wc -l blast/20210605-cgigas-roslin-mito-blastx.outfmt6

# Convert pipes to tab to isolate Uniprot accession
tr '|' '\t' < blast/20210605-cgigas-roslin-mito-blastx.outfmt6 \
> blast/cgigas-roslin-mito-blastx.outfmt6.codeIsolated

# Extract column with transcript ID
# Convert _ to tab
# Extract column with transcript ID
# Add XM_ to the front of each ID
cut -f2 blast/cgigas-roslin-mito-blastx.outfmt6.codeIsolated \
| tr "_" "\t" \
| cut -f5 \
| sed 's/^/XM_/' \
> blast/cgigas-roslin-mito-blastx.outfmt6.transcriptID
# head cgigas-roslin-mito-blastx.outfmt6.transcriptID
wc -l blast/cgigas-roslin-mito-blastx.outfmt6.transcriptID

# Paste original transcript ID with codeIsolated file
# Check output
paste blast/cgigas-roslin-mito-blastx.outfmt6.transcriptID blast/cgigas-roslin-mito-blastx.outfmt6.codeIsolated \
> blast/cgigas-roslin-mito-blastx.outfmt6.codeIsolated.transcriptID
# head cgigas-roslin-mito-blastx.outfmt6.codeIsolated.transcriptID

# Reduce the number of columns using awk: accession code, transcript ID, and e-value
# Sort, and save as a new file.
awk -v OFS='\t' '{print $5, $1, $15}' < blast/cgigas-roslin-mito-blastx.outfmt6.codeIsolated.transcriptID | sort \
> blast/gigas-blast-sort.tab
#head gigas-blast-sort.tab

# Download Uniprot database with GOterm information
#head uniprot-SP-GO.sorted
wc -l blast/uniprot-SP-GO.sorted

# Join the first column in the first file with the first column in the second file
# The files are tab delimited, and the output should also be tab delimited (-t $'\t')
join -1 1 -2 1 -t $'\t' \
blast/gigas-blast-sort.tab \
blast/uniprot-SP-GO.sorted \
> blast/gigas-blast-annot.tab
#head gigas-blast-annot.tab
wc -l blast/gigas-blast-annot.tab

# Extract columns 1 (accession), 2 (transcript ID), and 14 (GOterms)
# Save output
cut -f1,2,14 blast/gigas-blast-annot.tab \
> blast/_blast-annot.tab
# head _blast-annot.tab
wc -l blast/_blast-annot.tab

# Extract columns 2 and 3 to create list needed for topGO gene enrichment
# Convert ; to , for topGO formatting
cut -f2,3 blast/_blast-annot.tab \
| tr ";" "," \
> blast/geneid2go.tab
head blast/geneid2go.tab

```

```{bash, engine.path="C:/Windows/System32/bash.exe"}
#Isolate column with ID information
#Convert multiple delimiters to tabs
#Isolate transcript and gene ID columns
#Ensure they are tab-delimited
#Sort by gene ID
#Save output

mkdir GOterm_annotation/

cut -f9 GFF/cgigas_uk_roslin_v1_mRNA.gff \
| tr "=;:,-" "\t" \
| cut -f3,9 \
| awk '{print $2"\t"$1}' \
| sort \
> blast/cgigas_uk_roslin_v1_mRNA.transcriptID.geneID

```

# Modify blast output
```{bash, engine.path="C:/Windows/System32/bash.exe"}
# This script was originally written to address a specific problem that Rhonda was having

cd blast/
# input_file is the initial, "problem" file
# file is an intermediate file that most of the program works upon
# output_file is the final file produced by the script
input_file="_blast-annot.tab"
file="_intermediate.file"
output_file="_blast-GO-unfolded.tab"

# sed command substitutes the "; " sequence to a tab and writes the new format to a new file.
# This character sequence is how the GO terms are delimited in their field.
sed $'s/; /\t/g' "$input_file" > "$file"

# Identify first field containing a GO term.
# Search file with grep for "GO:" and pipe to awk.
# Awk sets tab as field delimiter (-F'\t'), runs a for loop that looks for "GO:" (~/GO:/), and then prints the field number).
# Awk results are piped to sort, which sorts unique by number (-ug).
# Sort results are piped to head to retrieve the lowest value (i.e. the top of the list; "-n1").
begin_goterms=$(grep "GO:" "$file" | awk -F'\t' '{for (i=1;i<=NF;i++) if($i ~/GO:/) print i}' | sort -ug | head -n1)

# While loop to process each line of the input file.
while read -r line
	do
	
	# Send contents of the current line to awk.
	# Set the field separator as a tab (-F'\t') and print the number of fields in that line.
	# Save the results of the echo/awk pipe (i.e. number of fields) to the variable "max_field".
	max_field=$(echo "$line" | awk -F'\t' '{print NF}')

	# Send contents of current line to cut.
	# Cut fields (i.e. retain those fields) 1-12.
	# Save the results of the echo/cut pipe (i.e. fields 1-12) to the variable "fixed_fields"
	fixed_fields=$(echo "$line" | cut -f1-2)

	# Since not all the lines contain the same number of fields (e.g. may not have GO terms),
	# evaluate the number of fields in each line to determine how to handle current line.

	# If the value in max_field is less than the field number where the GO terms begin,
	# then just print the current line (%s) followed by a newline (\n).
	if (( "$max_field" < "$begin_goterms" ))
		then printf "%s\n" "$line"
			else

			# Send contents of current line (which contains GO terms) to cut.
			# Cut fields (i.e. retain those fields) 13 to whatever the last field is in the curent line.
			# Save the results of the echo/cut pipe (i.e. all the GO terms fields) to the variable "goterms".
			goterms=$(echo "$line" | cut -f"$begin_goterms"-"$max_field")
			
			# Assign values in the variable "goterms" to a new indexed array (called "array"), 
			# with tab delimiter (IFS=$'\t')
			IFS=$'\t' read -r -a array <<<"$goterms"
			
			# Iterate through each element of the array.
			# Print the first 12 fields (i.e. the fields stored in "fixed_fields") followed by a tab (%s\t).
			# Print the current element in the array (i.e. the current GO term) followed by a new line (%s\n).
			for element in "${!array[@]}"	
				do printf "%s\t%s\n" "$fixed_fields" "${array[$element]}"
			done
	fi

# Send the input file into the while loop and send the output to a file named "rhonda_fixed.txt".
done < "$file" > "$output_file"
```

# Generate GOslim library (install in bash: conda install -c bioconda gsort)
```{bash, engine.path="C:/Windows/System32/bash.exe"}

pwd

eval "$(/home/mattgeorgephd/miniconda3/bin/conda shell.bash hook)"
conda activate

eval "$(/usr/bin/awk shell.bash hook)"
awk activate

cd blast/
#Reorganize and sort columns
awk '{print $3"\t"$2}' _blast-GO-unfolded.tab | gsort -V > _blast-GO-unfolded.sorted

#Join files to get GOslim for each query (with duplicate GOslim / query removed)
join -1 1 -2 1 -t $'\t' \
_blast-GO-unfolded.sorted \
GO-GOslim.sorted \
| uniq | awk -F'\t' -v OFS='\t' '{print $2, $1, $3, $4, $5}' \
| sort \
> Blastquery-GOslim.tab
head Blastquery-GOslim.tab
wc -l Blastquery-GOslim.tab
```

# Generate Conversion Tables
```{r}
# GOslim terms
Blastquery_GOslim <- read.delim("blast/Blastquery-GOslim.tab", sep = "\t", header = FALSE)
colnames(Blastquery_GOslim) <- c('product_accession','GOID','GOterm','GOSlim','GOcat')

# Load feature_table
gigas_feature_table        <- read.delim("data/GCF_902806645.1_cgigas_uk_roslin_v1_feature_table.txt", sep = "\t", header = TRUE)
gigas_feature_table_mRNA   <- gigas_feature_table[gigas_feature_table$X..feature == "mRNA",]
gigas_feature_table_mRNA   <- gigas_feature_table_mRNA[,c("product_accession","GeneID","genomic_accession","start","end","symbol")]
colnames(gigas_feature_table_mRNA) <- c("product_accession","GeneID","genomic_accession","start","end","gene")

# Generate Master mRNA table w/ GOterms
gigas_feature_table_mRNA <- left_join(gigas_feature_table_mRNA, Blastquery_GOslim, by = 'product_accession')

```

# remove commas in gene "description" to prevent errors when saving with write.table w/ "," seperators
```{r}
control_ploidy$description       <- gsub(",", ";", control_ploidy$description)
heat_ploidy$description          <- gsub(",", ";", heat_ploidy$description)
desiccation_ploidy$description   <- gsub(",", ";", desiccation_ploidy$description)
diploid_heat$description         <- gsub(",", ";", diploid_heat$description)
diploid_desiccation$description  <- gsub(",", ";", diploid_desiccation$description)
triploid_heat$description        <- gsub(",", ";", triploid_heat$description)
triploid_desiccation$description <- gsub(",", ";", triploid_desiccation$description)

control_ploidy_bg$description       <- gsub(",", ";", control_ploidy_bg$description)
heat_ploidy_bg$description          <- gsub(",", ";", heat_ploidy_bg$description)
desiccation_ploidy_bg$description   <- gsub(",", ";", desiccation_ploidy_bg$description)
diploid_heat_bg$description         <- gsub(",", ";", diploid_heat_bg$description)
diploid_desiccation_bg$description  <- gsub(",", ";", diploid_desiccation_bg$description)
triploid_heat_bg$description        <- gsub(",", ";", triploid_heat_bg$description)
triploid_desiccation_bg$description <- gsub(",", ";", triploid_desiccation_bg$description)

diploid_MS_unique$description <- gsub(",", ";", diploid_MS_unique$description)
triploid_MS_unique$description <- gsub(",", ";", triploid_MS_unique$description)

```

# Count up and down regulated genes, generate table
```{r}
direction <- data.frame('comparison' = c('control_ploidy'        ,
                                         'heat_ploidy'           ,
                                         'desiccation_ploidy'    ,
                                         'diploid_heat'          ,
                                         'diploid_desiccation'   ,
                                         'triploid_heat'         ,
                                         'triploid_desiccation'  ,
                                         'diploid_MS_unique'     ,
                                         'triploid_MS_unique') , 
                        'total'    = 0, 
                        'positive' = 0 )

direction$total[1]    <- length(control_ploidy$log2FoldChange)
direction$positive[1] <- length(control_ploidy$log2FoldChange[control_ploidy$log2FoldChange>0])

direction$total[2]    <- length(heat_ploidy$log2FoldChange)
direction$positive[2] <- length(heat_ploidy$log2FoldChange[heat_ploidy$log2FoldChange>0])

direction$total[3]    <- length(desiccation_ploidy$log2FoldChange)
direction$positive[3] <- length(desiccation_ploidy$log2FoldChange[desiccation_ploidy$log2FoldChange>0])

direction$total[4]    <- length(diploid_heat$log2FoldChange)
direction$positive[4] <- length(diploid_heat$log2FoldChange[diploid_heat$log2FoldChange>0])

direction$total[5]    <- length(diploid_desiccation$log2FoldChange)
direction$positive[5] <- length(diploid_desiccation$log2FoldChange[diploid_desiccation$log2FoldChange>0])

direction$total[6]    <- length(triploid_heat$log2FoldChange)
direction$positive[6] <- length(triploid_heat$log2FoldChange[triploid_heat$log2FoldChange>0])

direction$total[7]    <- length(triploid_desiccation$log2FoldChange)
direction$positive[7] <- length(triploid_desiccation$log2FoldChange[triploid_desiccation$log2FoldChange>0])

direction$total[8]    <- length(diploid_MS_unique$log2FoldChange)
direction$positive[8] <- length(diploid_MS_unique$log2FoldChange[diploid_MS_unique$log2FoldChange>0])

direction$total[9]    <- length(triploid_MS_unique$log2FoldChange)
direction$positive[9] <- length(triploid_MS_unique$log2FoldChange[triploid_MS_unique$log2FoldChange>0])

direction$negative         <- direction$total - direction$positive
direction$percent_negative <- round(direction$negative/direction$total*100,digits = 1)
direction$percent_positive <- round(direction$positive/direction$total*100,digits = 1)

write.table(direction, file = "gene_tables/DEG_direction.txt", sep = ",", quote = FALSE, row.names = FALSE)

```

# Separate genes (based on LOCID) w/ uncharactered proteins, run stats
```{r}

grx <- glob2rx("unchar*")

control_ploidy_NA              <- with(control_ploidy,       subset(control_ploidy,       subset = grepl(glob2rx("LOCNA*"), gene)))
control_ploidy                 <- anti_join(control_ploidy, control_ploidy_NA, by = 'gene')
control_ploidy_uncharacterized <- with(control_ploidy, subset(control_ploidy, subset = grepl(grx, description)))
control_ploidy_characterized   <- anti_join(control_ploidy, control_ploidy_uncharacterized, by = 'gene')

heat_ploidy_NA              <- with(heat_ploidy,          subset(heat_ploidy,          subset = grepl(glob2rx("LOCNA*"), gene)))
heat_ploidy                 <- anti_join(heat_ploidy, heat_ploidy_NA, by = 'gene')
heat_ploidy_uncharacterized <- with(heat_ploidy, subset(heat_ploidy, subset = grepl(grx, description)))
heat_ploidy_characterized   <- anti_join(heat_ploidy, heat_ploidy_uncharacterized, by = 'gene')

desiccation_ploidy_NA              <- with(desiccation_ploidy,   subset(desiccation_ploidy,   subset = grepl(glob2rx("LOCNA*"), gene)))
desiccation_ploidy                 <- anti_join(desiccation_ploidy, desiccation_ploidy_NA, by = 'gene')
desiccation_ploidy_uncharacterized <- with(desiccation_ploidy, subset(desiccation_ploidy, subset = grepl(grx, description)))
desiccation_ploidy_characterized   <- anti_join(desiccation_ploidy, desiccation_ploidy_uncharacterized, by = 'gene')

diploid_heat_NA              <- with(diploid_heat ,        subset(diploid_heat,         subset = grepl(glob2rx("LOCNA*"), gene)))
diploid_heat                 <- anti_join(diploid_heat, diploid_heat_NA, by = 'gene')
diploid_heat_uncharacterized <- with(diploid_heat, subset(diploid_heat, subset = grepl(grx, description)))
diploid_heat_characterized   <- anti_join(diploid_heat, diploid_heat_uncharacterized, by = 'gene')

diploid_desiccation_NA              <- with(diploid_desiccation,  subset(diploid_desiccation,  subset = grepl(glob2rx("LOCNA*"), gene)))
diploid_desiccation                 <- anti_join(diploid_desiccation, diploid_desiccation_NA, by = 'gene')
diploid_desiccation_uncharacterized <- with(diploid_desiccation, subset(diploid_desiccation, subset = grepl(grx, description)))
diploid_desiccation_characterized   <- anti_join(diploid_desiccation, diploid_desiccation_uncharacterized, by = 'gene')

triploid_heat_NA              <- with(triploid_heat,        subset(triploid_heat,        subset = grepl(glob2rx("LOCNA*"), gene)))
triploid_heat                 <- anti_join(triploid_heat, triploid_heat_NA, by = 'gene')
triploid_heat_uncharacterized <- with(triploid_heat, subset(triploid_heat, subset = grepl(grx, description)))
triploid_heat_characterized   <- anti_join(triploid_heat, triploid_heat_uncharacterized, by = 'gene')

triploid_desiccation_NA              <- with(triploid_desiccation, subset(triploid_desiccation, subset = grepl(glob2rx("LOCNA*"), gene)))
triploid_desiccation                 <- anti_join(triploid_desiccation, triploid_desiccation_NA, by = 'gene')
triploid_desiccation_uncharacterized <- with(triploid_desiccation, subset(triploid_desiccation, subset = grepl(grx, description)))
triploid_desiccation_characterized   <- anti_join(triploid_desiccation, triploid_desiccation_uncharacterized, by = 'gene')

diploid_MS_unique_NA                 <- with(diploid_MS_unique, subset(diploid_MS_unique, subset = grepl(glob2rx("LOCNA*"), gene)))
diploid_MS_unique                    <- anti_join(diploid_MS_unique, diploid_MS_unique_NA, by = 'gene')
diploid_MS_unique_uncharacterized    <- with(diploid_MS_unique, subset(diploid_MS_unique, subset = grepl(grx, description)))
diploid_MS_unique_characterized      <- anti_join(diploid_MS_unique, diploid_MS_unique_uncharacterized, by = 'gene')

triploid_MS_unique_NA                 <- with(triploid_MS_unique, subset(triploid_MS_unique, subset = grepl(glob2rx("LOCNA*"), gene)))
triploid_MS_unique                    <- anti_join(triploid_MS_unique, triploid_MS_unique_NA, by = 'gene')
triploid_MS_unique_uncharacterized    <- with(triploid_MS_unique, subset(triploid_MS_unique, subset = grepl(grx, description)))
triploid_MS_unique_characterized      <- anti_join(triploid_MS_unique, triploid_MS_unique_uncharacterized, by = 'gene')

# Count number of uncharacterized genes
uncharacterized <- data.frame(known = rep(0, 9), 
                              unknown = rep(0, 9), 
                              percent_unknown = rep(0, 9))

uncharacterized$known <- c(nrow(control_ploidy_characterized),
                           nrow(heat_ploidy_characterized),
                           nrow(desiccation_ploidy_characterized),
                           nrow(diploid_heat_characterized),
                           nrow(diploid_desiccation_characterized),
                           nrow(triploid_heat_characterized),
                           nrow(triploid_desiccation_characterized),
                           nrow(diploid_MS_unique_characterized),
                           nrow(triploid_MS_unique_characterized))

uncharacterized$unknown <- c(nrow(control_ploidy_uncharacterized),
                           nrow(heat_ploidy_uncharacterized),
                           nrow(desiccation_ploidy_uncharacterized),
                           nrow(diploid_heat_uncharacterized),
                           nrow(diploid_desiccation_uncharacterized),
                           nrow(triploid_heat_uncharacterized),
                           nrow(triploid_desiccation_uncharacterized),
                           nrow(diploid_MS_unique_uncharacterized),
                           nrow(triploid_MS_unique_uncharacterized))

uncharacterized$percent_unknown <- (uncharacterized$unknown / (uncharacterized$unknown + uncharacterized$known))

row.names(uncharacterized) <- c('control_ploidy',
                                'heat_ploidy', 
                                'desiccation_ploidy', 
                                'diploid_heat', 
                                'diploid_desiccation', 
                                'triploid_heat', 
                                'triploid_desiccation',
                                'diploid_MS_unique',
                                'triploid_MS_unique')

write.table(uncharacterized, file = "gene_tables/uncharacterized_stats.txt", sep = ",", quote = FALSE, row.names = TRUE)

```

# Remove uncharacterized genes
```{r}
grx <- glob2rx("unchar*")

control_ploidy_bg_NA              <- with(control_ploidy_bg,       subset(control_ploidy_bg,       subset = grepl(glob2rx("LOCNA*"), gene)))
control_ploidy_bg                 <- anti_join(control_ploidy_bg, control_ploidy_bg_NA, by = 'gene')
control_ploidy_bg_uncharacterized <- with(control_ploidy_bg, subset(control_ploidy_bg, subset = grepl(grx, description)))
control_ploidy_bg_characterized   <- anti_join(control_ploidy_bg, control_ploidy_bg_uncharacterized, by = 'gene')

heat_ploidy_bg_NA              <- with(heat_ploidy_bg,          subset(heat_ploidy_bg,          subset = grepl(glob2rx("LOCNA*"), gene)))
heat_ploidy_bg                 <- anti_join(heat_ploidy_bg, heat_ploidy_bg_NA, by = 'gene')
heat_ploidy_bg_uncharacterized <- with(heat_ploidy_bg, subset(heat_ploidy_bg, subset = grepl(grx, description)))
heat_ploidy_bg_characterized   <- anti_join(heat_ploidy_bg, heat_ploidy_bg_uncharacterized, by = 'gene')

desiccation_ploidy_bg_NA              <- with(desiccation_ploidy_bg,   subset(desiccation_ploidy_bg,   subset = grepl(glob2rx("LOCNA*"), gene)))
desiccation_ploidy_bg                 <- anti_join(desiccation_ploidy_bg, desiccation_ploidy_bg_NA, by = 'gene')
desiccation_ploidy_bg_uncharacterized <- with(desiccation_ploidy_bg, subset(desiccation_ploidy_bg, subset = grepl(grx, description)))
desiccation_ploidy_bg_characterized   <- anti_join(desiccation_ploidy_bg, desiccation_ploidy_bg_uncharacterized, by = 'gene')

diploid_heat_bg_NA              <- with(diploid_heat_bg ,        subset(diploid_heat_bg,         subset = grepl(glob2rx("LOCNA*"), gene)))
diploid_heat_bg                 <- anti_join(diploid_heat_bg, diploid_heat_bg_NA, by = 'gene')
diploid_heat_bg_uncharacterized <- with(diploid_heat_bg, subset(diploid_heat_bg, subset = grepl(grx, description)))
diploid_heat_bg_characterized   <- anti_join(diploid_heat_bg, diploid_heat_bg_uncharacterized, by = 'gene')

diploid_desiccation_bg_NA              <- with(diploid_desiccation_bg,  subset(diploid_desiccation_bg,  subset = grepl(glob2rx("LOCNA*"), gene)))
diploid_desiccation_bg                 <- anti_join(diploid_desiccation_bg, diploid_desiccation_bg_NA, by = 'gene')
diploid_desiccation_bg_uncharacterized <- with(diploid_desiccation_bg, subset(diploid_desiccation_bg, subset = grepl(grx, description)))
diploid_desiccation_bg_characterized   <- anti_join(diploid_desiccation_bg, diploid_desiccation_bg_uncharacterized, by = 'gene')

triploid_heat_bg_NA              <- with(triploid_heat_bg,        subset(triploid_heat_bg,        subset = grepl(glob2rx("LOCNA*"), gene)))
triploid_heat_bg                 <- anti_join(triploid_heat_bg, triploid_heat_bg_NA, by = 'gene')
triploid_heat_bg_uncharacterized <- with(triploid_heat_bg, subset(triploid_heat_bg, subset = grepl(grx, description)))
triploid_heat_bg_characterized   <- anti_join(triploid_heat_bg, triploid_heat_bg_uncharacterized, by = 'gene')

triploid_desiccation_bg_NA              <- with(triploid_desiccation_bg, subset(triploid_desiccation_bg, subset = grepl(glob2rx("LOCNA*"), gene)))
triploid_desiccation_bg                 <- anti_join(triploid_desiccation_bg, triploid_desiccation_bg_NA, by = 'gene')
triploid_desiccation_bg_uncharacterized <- with(triploid_desiccation_bg, subset(triploid_desiccation_bg, subset = grepl(grx, description)))
triploid_desiccation_bg_characterized   <- anti_join(triploid_desiccation_bg, triploid_desiccation_bg_uncharacterized, by = 'gene')

```

# Match GO terms w/ characterized GeneIDs
```{r}
# Desired file structure from TREATMENT.GeneIDs.geneOverlap.transcriptIDs.GOAnnot file:
# "transcript", "geneID", "chr", "start", "end","meth.diff","GOID", "GOterm", "GOSlim", "GOcat"

control_ploidy_characterized        <- left_join(control_ploidy_characterized       , gigas_feature_table_mRNA, by = 'gene')
heat_ploidy_characterized           <- left_join(heat_ploidy_characterized          , gigas_feature_table_mRNA, by = 'gene')
desiccation_ploidy_characterized    <- left_join(desiccation_ploidy_characterized   , gigas_feature_table_mRNA, by = 'gene')
diploid_heat_characterized          <- left_join(diploid_heat_characterized         , gigas_feature_table_mRNA, by = 'gene')
diploid_desiccation_characterized   <- left_join(diploid_desiccation_characterized  , gigas_feature_table_mRNA, by = 'gene')
triploid_heat_characterized         <- left_join(triploid_heat_characterized        , gigas_feature_table_mRNA, by = 'gene')
triploid_desiccation_characterized  <- left_join(triploid_desiccation_characterized , gigas_feature_table_mRNA, by = 'gene')
diploid_MS_unique_characterized     <- left_join(diploid_MS_unique_characterized    , gigas_feature_table_mRNA, by = 'gene')
triploid_MS_unique_characterized    <- left_join(triploid_MS_unique_characterized   , gigas_feature_table_mRNA, by = 'gene')

control_ploidy        <- left_join(control_ploidy       , gigas_feature_table_mRNA, by = 'gene')
heat_ploidy           <- left_join(heat_ploidy          , gigas_feature_table_mRNA, by = 'gene')
desiccation_ploidy    <- left_join(desiccation_ploidy   , gigas_feature_table_mRNA, by = 'gene')
diploid_heat          <- left_join(diploid_heat         , gigas_feature_table_mRNA, by = 'gene')
diploid_desiccation   <- left_join(diploid_desiccation  , gigas_feature_table_mRNA, by = 'gene')
triploid_heat         <- left_join(triploid_heat        , gigas_feature_table_mRNA, by = 'gene')
triploid_desiccation  <- left_join(triploid_desiccation , gigas_feature_table_mRNA, by = 'gene')
diploid_MS_unique     <- left_join(diploid_MS_unique    , gigas_feature_table_mRNA, by = 'gene')
triploid_MS_unique    <- left_join(triploid_MS_unique , gigas_feature_table_mRNA, by = 'gene')

control_ploidy_bg_characterized        <- left_join(control_ploidy_bg_characterized       , gigas_feature_table_mRNA, by = 'gene')
heat_ploidy_bg_characterized           <- left_join(heat_ploidy_bg_characterized          , gigas_feature_table_mRNA, by = 'gene')
desiccation_bg_characterized           <- left_join(desiccation_ploidy_bg_characterized   , gigas_feature_table_mRNA, by = 'gene')
diploid_heat_bg_characterized          <- left_join(diploid_heat_bg_characterized         , gigas_feature_table_mRNA, by = 'gene')
diploid_desiccation_bg_characterized   <- left_join(diploid_desiccation_bg_characterized  , gigas_feature_table_mRNA, by = 'gene')
triploid_heat_bg_characterized         <- left_join(triploid_heat_bg_characterized        , gigas_feature_table_mRNA, by = 'gene')
triploid_desiccation_bg_characterized  <- left_join(triploid_desiccation_bg_characterized , gigas_feature_table_mRNA, by = 'gene')

control_ploidy_bg        <- left_join(control_ploidy_bg       , gigas_feature_table_mRNA, by = 'gene')
heat_ploidy_bg           <- left_join(heat_ploidy_bg          , gigas_feature_table_mRNA, by = 'gene')
desiccation_ploidy_bg    <- left_join(desiccation_ploidy_bg   , gigas_feature_table_mRNA, by = 'gene')
diploid_heat_bg          <- left_join(diploid_heat_bg         , gigas_feature_table_mRNA, by = 'gene')
diploid_desiccation_bg   <- left_join(diploid_desiccation_bg  , gigas_feature_table_mRNA, by = 'gene')
triploid_heat_bg         <- left_join(triploid_heat_bg        , gigas_feature_table_mRNA, by = 'gene')
triploid_desiccation_bg  <- left_join(triploid_desiccation_bg , gigas_feature_table_mRNA, by = 'gene')


```

# Remove genes w/ NA
```{r}

control_ploidy              <- control_ploidy[complete.cases(control_ploidy),]
heat_ploidy                 <- heat_ploidy[complete.cases(heat_ploidy),]
desiccation_ploidy          <- desiccation_ploidy[complete.cases(desiccation_ploidy),]
diploid_heat                <- diploid_heat[complete.cases(diploid_heat),]
diploid_desiccation         <- diploid_desiccation[complete.cases(diploid_desiccation),]
triploid_heat               <- triploid_heat[complete.cases(triploid_heat),]
triploid_desiccation        <- triploid_desiccation[complete.cases(triploid_desiccation),]
diploid_MS_unique           <- diploid_MS_unique[complete.cases(diploid_MS_unique),]
triploid_MS_unique          <- triploid_MS_unique[complete.cases(triploid_MS_unique),]

control_ploidy_bg              <- control_ploidy_bg[complete.cases(control_ploidy_bg),]
heat_ploidy_bg                 <- heat_ploidy_bg[complete.cases(heat_ploidy_bg),]
desiccation_ploidy_bg          <- desiccation_ploidy_bg[complete.cases(desiccation_ploidy_bg),]
diploid_heat_bg                <- diploid_heat_bg[complete.cases(diploid_heat_bg),]
diploid_desiccation_bg         <- diploid_desiccation_bg[complete.cases(diploid_desiccation_bg),]
triploid_heat_bg               <- triploid_heat_bg[complete.cases(triploid_heat_bg),]
triploid_desiccation_bg        <- triploid_desiccation_bg[complete.cases(triploid_desiccation_bg),]

```

# Write GOterm tables
```{r}

write.table(control_ploidy,       file = "gene_tables/control_ploidy.txt",       sep = "\t", quote = FALSE, row.names = F)
write.table(heat_ploidy,          file = "gene_tables/heat_ploidy.txt",          sep = "\t", quote = FALSE, row.names = F)
write.table(desiccation_ploidy,   file = "gene_tables/desiccation_ploidy.txt",   sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_heat,         file = "gene_tables/diploid_heat.txt",         sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_desiccation,  file = "gene_tables/diploid_desiccation.txt",  sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_heat,        file = "gene_tables/triploid_heat.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_desiccation, file = "gene_tables/triploid_desiccation.txt", sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_MS_unique,    file = "gene_tables/diploid_MS_unique.txt", sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_MS_unique,   file = "gene_tables/triploid_MS_unique.txt", sep = "\t", quote = FALSE, row.names = F)

write.table(control_ploidy_characterized,       file = "gene_tables/control_ploidy_characterized.txt",       sep = "\t", quote = FALSE, row.names = F)
write.table(heat_ploidy_characterized,          file = "gene_tables/heat_ploidy_characterized.txt",          sep = "\t", quote = FALSE, row.names = F)
write.table(desiccation_ploidy_characterized,   file = "gene_tables/desiccation_ploidy_characterized.txt",   sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_heat_characterized,         file = "gene_tables/diploid_heat_characterized.txt",         sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_desiccation_characterized,  file = "gene_tables/diploid_desiccation_characterized.txt",  sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_heat_characterized,        file = "gene_tables/triploid_heat_characterized.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_desiccation_characterized, file = "gene_tables/triploid_desiccation_characterized.txt", sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_MS_unique_characterized,        file = "gene_tables/diploid_MS_unique_characterized.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_MS_unique_characterized, file = "gene_tables/triploid_MS_unique_characterized.txt", sep = "\t", quote = FALSE, row.names = F)

write.table(control_ploidy_uncharacterized,       file = "gene_tables/control_ploidy_uncharacterized.txt",       sep = "\t", quote = FALSE, row.names = F)
write.table(heat_ploidy_uncharacterized,          file = "gene_tables/heat_ploidy_uncharacterized.txt",          sep = "\t", quote = FALSE, row.names = F)
write.table(desiccation_ploidy_uncharacterized,   file = "gene_tables/desiccation_ploidy_uncharacterized.txt",   sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_heat_uncharacterized,         file = "gene_tables/diploid_heat_uncharacterized.txt",         sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_desiccation_uncharacterized,  file = "gene_tables/diploid_desiccation_uncharacterized.txt",  sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_heat_uncharacterized,        file = "gene_tables/triploid_heat_uncharacterized.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_desiccation_uncharacterized, file = "gene_tables/triploid_desiccation_uncharacterized.txt", sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_MS_unique_uncharacterized,    file = "gene_tables/diploid_MS_unique_uncharacterized.txt",    sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_MS_unique_uncharacterized, file = "gene_tables/triploid_MS_unique_uncharacterized.txt",     sep = "\t", quote = FALSE, row.names = F)

write.table(control_ploidy$gene,       file = "gene_tables/control_ploidy_all_LOCID.txt",      sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(heat_ploidy$gene,          file = "gene_tables/heat_ploidy_all_LOCID.txt",         sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(desiccation_ploidy$gene,   file = "gene_tables/desiccation_ploidy_all_LOCID.txt",  sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(diploid_heat$gene,         file = "gene_tables/diploid_heat_all_LOCID.txt",  sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(diploid_desiccation$gene,  file = "gene_tables/diploid_desiccation_all_LOCID.txt",  sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(triploid_heat$gene,        file = "gene_tables/triploid_heat_all_LOCID.txt",        sep = "\t", quote = FALSE, row.names = F, col.names = F)
write.table(triploid_desiccation$gene, file = "gene_tables/triploid_desiccation_all_LOCID.txt", sep = "\t", quote = FALSE, row.names = F, col.names = F)

write.table(control_ploidy_bg_characterized,       file = "gene_tables/control_ploidy_bg_characterized.txt",       sep = "\t", quote = FALSE, row.names = F)
write.table(heat_ploidy_bg_characterized,          file = "gene_tables/heat_ploidy_bg_characterized.txt",          sep = "\t", quote = FALSE, row.names = F)
write.table(desiccation_ploidy_bg_characterized,   file = "gene_tables/desiccation_ploidy_bg_characterized.txt",   sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_heat_bg_characterized,         file = "gene_tables/diploid_heat_bg_characterized.txt",         sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_desiccation_bg_characterized,  file = "gene_tables/diploid_desiccation_bg_characterized.txt",  sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_heat_bg_characterized,        file = "gene_tables/triploid_heat_bg_characterized.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_desiccation_bg_characterized, file = "gene_tables/triploid_desiccation_bg_characterized.txt", sep = "\t", quote = FALSE, row.names = F)

write.table(control_ploidy_bg,       file = "gene_tables/control_ploidy_bg.txt",       sep = "\t", quote = FALSE, row.names = F)
write.table(heat_ploidy_bg,          file = "gene_tables/heat_ploidy_bg.txt",          sep = "\t", quote = FALSE, row.names = F)
write.table(desiccation_ploidy_bg,   file = "gene_tables/desiccation_ploidy_bg.txt",   sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_heat_bg,         file = "gene_tables/diploid_heat_bg.txt",         sep = "\t", quote = FALSE, row.names = F)
write.table(diploid_desiccation_bg,  file = "gene_tables/diploid_desiccation_bg.txt",  sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_heat_bg,        file = "gene_tables/triploid_heat_bg.txt",        sep = "\t", quote = FALSE, row.names = F)
write.table(triploid_desiccation_bg, file = "gene_tables/triploid_desiccation_bg.txt", sep = "\t", quote = FALSE, row.names = F)

beep(2); Sys.sleep(0.3); beep(2); Sys.sleep(0.3); beep(2); Sys.sleep(0.3); beep(4)
```


